### ISSUE TITLE
alternator: write failure when GSI has two new key columns and one has no value

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#6008]
When an Alternator GSI (i.e., materialized view) has a regular-base-column X as a key column in the view, if one inserts an item **without** a value for X into the base table, this is fine - but the item will not be inserted into the view. We already have a test case for this: `test_gsi.py::test_gsi_missing_attribute_3`.

However, Alternator also allows a GSI to have *two* regular base columns, say X and Y, as the key columns (one as the HASH key the second the RANGE key). In this case, if either X or Y (or both) don't have a value in an item - the item shouldn't appear in the view. But this case does **not** work correctly, as discovered by a user who tried this case. The client gets an error for the write:

```
Internal server error: exceptions::mutation_write_failure_exception (Operation failed for alternator_alternator_Test_1583967177951.alternator_Test_1583967177951 - received 0 responses and 1 failures from 1 CL=LOCAL_QUORUM.)
```

The log contains a bit more information:
```
ERROR 2020-03-12 00:52:59,841 [shard 2] storage_proxy - exception during mutation write to 127.0.0.1: seastar::internal::backtraced<std::out_of_range> (Column not found for id = 1 Backtrace: ....
```

The following test reproduces this bug. It succeeds on AWS, but fails with the above write errors and log messages on Alternator:

```python
# Test that when a table has a GSI, if the indexed attribute is missing, the
# item is added to the base table but not the index.
# This is the same feature we already tested in test_gsi_missing_attribute()
# above; But in that test we used test_table_gsi_2, which had just one indexed
# attribute. In the test here we use test_table_gsi_3 which has two attributes
# being indexed, and more possibilities of which might be missing - if one of
# the two is missing, the item should not be indexed
def test_gsi_missing_attribute_3(test_table_gsi_3):
    p = random_string()
    a = random_string()
    b = random_string()
    # First, add an item with a missing "a" value. It should appear in the
    # base table, but not in the index:
    test_table_gsi_3.put_item(Item={'p':  p, 'b': b})
    assert test_table_gsi_3.get_item(Key={'p':  p})['Item'] == {'p': p, 'b': b}
    # Note: with eventually consistent read, we can't really be sure that
    # an item will "never" appear in the index. We hope that if a bug exists
    # and such an item did appear, sometimes the delay here will be enough
    # for the unexpected item to become visible.
    assert not any([i['p'] == p for i in full_scan(test_table_gsi_3, IndexName='hello')])
    # Same thing for an item with a missing "b" value:
    test_table_gsi_3.put_item(Item={'p':  p, 'a': a})
    assert test_table_gsi_3.get_item(Key={'p':  p})['Item'] == {'p': p, 'a': a}
    assert not any([i['p'] == p for i in full_scan(test_table_gsi_3, IndexName='hello')])
    # And an item missing both:
    test_table_gsi_3.put_item(Item={'p':  p})
    assert test_table_gsi_3.get_item(Key={'p':  p})['Item'] == {'p': p}
    assert not any([i['p'] == p for i in full_scan(test_table_gsi_3, IndexName='hello')])
```

**Comment by nyh** [comment]
Sent to the mailing list a patch with an xfailing test, `test_gsi.py::test_gsi_missing_attribute_3`, reproducing this issue (basically the same test shown above).

**Comment by nyh** [comment]
Patch set to the mailing list, titled "materialized views: fix corner case of view updates used by Alternator".

**Comment by nyh** [comment]
This commit is already in 4.0 so doesn't need backporting there.
Because this is a bug fix, backported it now to 3.3 as well, even though Alternator was experimental in 3.3.
